File: packages/android/src/main/java/com/kozmos/components/Timeline/Timeline.kt
package com.kozmos.components.timeline

import com.kozmos.tokens.KozmosDimensions

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import com.kozmos.tokens.KozmosThemeTokens

data class TimelineItem(
    val time: String,
    val title: String,
    val description: String
)

enum class KozmosTimelineDensity {
    Default,
    Compact
}

private val LocalKozmosTimelineDensity = staticCompositionLocalOf {
    KozmosTimelineDensity.Default
}

private val KozmosTimelineDensity.contentSpacing
    get() = if (this == KozmosTimelineDensity.Compact) {
        KozmosDimensions.primitivesLayoutSpacing25
    } else {
        KozmosDimensions.primitivesLayoutSpacing50
    }

private val KozmosTimelineDensity.itemBottomPadding
    get() = if (this == KozmosTimelineDensity.Compact) {
        KozmosDimensions.primitivesLayoutSpacing200
    } else {
        KozmosDimensions.primitivesLayoutSpacing400
    }

@Composable
fun KozmosTimeline(
    modifier: Modifier = Modifier,
    density: KozmosTimelineDensity = KozmosTimelineDensity.Default,
    content: @Composable ColumnScope.() -> Unit
) {
    CompositionLocalProvider(LocalKozmosTimelineDensity provides density) {
        Column(
            modifier = modifier,
            verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing0)
        ) {
            content()
        }
    }
}

@Composable
fun KozmosTimelineItem(
    modifier: Modifier = Modifier,
    isLast: Boolean = false,
    content: @Composable ColumnScope.() -> Unit
) {
    val density = LocalKozmosTimelineDensity.current

    Row(
        modifier = modifier
            .fillMaxWidth()
            .height(IntrinsicSize.Min)
    ) {
        Column(
            horizontalAlignment = androidx.compose.ui.Alignment.CenterHorizontally,
            modifier = Modifier.width(KozmosDimensions.primitivesLayoutSizing500)
        ) {
            Box(
                modifier = Modifier
                    .size(KozmosDimensions.primitivesLayoutSpacing150)
                    .clip(CircleShape)
                    .background(KozmosThemeTokens.primitivesColorsTheme500)
            )
            if (!isLast) {
                Box(
                    modifier = Modifier
                        .width(KozmosDimensions.primitivesLayoutSpacing25)
                        .fillMaxHeight()
                        .background(KozmosThemeTokens.primitivesColorsBackground300)
                        .padding(top = KozmosDimensions.primitivesLayoutSpacing50)
                )
            }
        }
        
        Column(
            modifier = Modifier
                .padding(
                    start = KozmosDimensions.primitivesLayoutSpacing100, 
                    bottom = if (isLast) KozmosDimensions.primitivesLayoutSpacing0 else density.itemBottomPadding
                )
                .fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(density.contentSpacing)
        ) {
            content()
        }
    }
}

@Composable
fun KozmosTimelineTime(
    time: String,
    modifier: Modifier = Modifier
) {
    Text(
        text = time,
        style = MaterialTheme.typography.labelMedium,
        color = KozmosThemeTokens.primitivesColorsForeground500,
        modifier = modifier
    )
}

@Composable
fun KozmosTimelineTitle(
    title: String,
    modifier: Modifier = Modifier
) {
    Text(
        text = title,
        style = MaterialTheme.typography.titleMedium,
        fontWeight = FontWeight.Bold,
        color = KozmosThemeTokens.primitivesColorsForeground100,
        modifier = modifier
    )
}

@Composable
fun KozmosTimelineDescription(
    description: String,
    modifier: Modifier = Modifier
) {
    Text(
        text = description,
        style = MaterialTheme.typography.bodyMedium,
        color = KozmosThemeTokens.primitivesColorsForeground500,
        modifier = modifier
    )
}

File: packages/ios/Sources/Components/Timeline/Timeline.swift
import SwiftUI

public enum KozmosTimelineDensity {
    case `default`
    case compact

    var contentSpacing: CGFloat {
        switch self {
        case .default:
            return KozmosDimensions.primitivesLayoutSpacing50
        case .compact:
            return KozmosDimensions.primitivesLayoutSpacing25
        }
    }

    var itemBottomPadding: CGFloat {
        switch self {
        case .default:
            return KozmosDimensions.primitivesLayoutSpacing400
        case .compact:
            return KozmosDimensions.primitivesLayoutSpacing200
        }
    }
}

private struct KozmosTimelineDensityKey: EnvironmentKey {
    static let defaultValue: KozmosTimelineDensity = .default
}

private extension EnvironmentValues {
    var kozmosTimelineDensity: KozmosTimelineDensity {
        get { self[KozmosTimelineDensityKey.self] }
        set { self[KozmosTimelineDensityKey.self] = newValue }
    }
}

public struct TimelineItem: Identifiable {
    public let id = UUID()
    public let time: String
    public let title: String
    public let description: String
    
    public init(time: String, title: String, description: String) {
        self.time = time
        self.title = title
        self.description = description
    }
}

public struct KozmosTimeline<Content: View>: View {
    let content: Content
    let density: KozmosTimelineDensity
    
    public init(
        density: KozmosTimelineDensity = .default,
        @ViewBuilder content: () -> Content
    ) {
        self.density = density
        self.content = content()
    }
    
    public var body: some View {
        VStack(alignment: .leading, spacing: KozmosDimensions.primitivesLayoutSpacing0) {
            content
        }
        .environment(\.kozmosTimelineDensity, density)
        .padding()
    }
}

public struct KozmosTimelineItem<Content: View>: View {
    @Environment(\.kozmosTimelineDensity) private var density

    let content: Content
    let isLast: Bool
    
    public init(isLast: Bool = false, @ViewBuilder content: () -> Content) {
        self.isLast = isLast
        self.content = content()
    }
    
    public var body: some View {
        HStack(alignment: .top, spacing: KozmosDimensions.primitivesLayoutSpacing200) {
            VStack(spacing: KozmosDimensions.primitivesLayoutSpacing0) {
                Circle()
                    .fill(KozmosColors.primitivesColorsTheme500)
                    .frame(width: 12, height: 12)
                    .background(
                        Circle()
                            .fill(KozmosColors.primitivesColorsTheme500.opacity(0.2))
                            .frame(width: KozmosDimensions.primitivesLayoutSizing300, height: KozmosDimensions.primitivesLayoutSizing300)
                    )
                
                if !isLast {
                    Rectangle()
                        .fill(KozmosColors.primitivesColorsBackground300)
                        .frame(width: 2)
                        .frame(maxHeight: .infinity)
                        .padding(.top, KozmosDimensions.primitivesLayoutSpacing50)
                }
            }
            .frame(width: KozmosDimensions.primitivesLayoutSizing300)
            
            VStack(alignment: .leading, spacing: density.contentSpacing) {
                content
            }
            .padding(.bottom, !isLast ? density.itemBottomPadding : KozmosDimensions.primitivesLayoutSpacing0)
        }
        .fixedSize(horizontal: false, vertical: true)
    }
}

public struct KozmosTimelineTime: View {
    let time: String
    
    public init(_ time: String) {
        self.time = time
    }
    
    public var body: some View {
        Text(time)
            .font(.caption)
            .foregroundColor(KozmosColors.primitivesColorsForeground500)
    }
}

public struct KozmosTimelineTitle: View {
    let title: String
    
    public init(_ title: String) {
        self.title = title
    }
    
    public var body: some View {
        Text(title)
            .font(.headline)
            .foregroundColor(KozmosColors.primitivesColorsForeground100)
    }
}

public struct KozmosTimelineDescription: View {
    let description: String
    
    public init(_ description: String) {
        self.description = description
    }
    
    public var body: some View {
        Text(description)
            .font(.body)
            .foregroundColor(KozmosColors.primitivesColorsForeground500)
            .fixedSize(horizontal: false, vertical: true)
    }
}
